### PR TITLE
make use of C++11's auto type deduction for temporary variables

### PR DESCRIPTION
**Pull request by heterodoxic**
This is just one of those tiny steps towards the goal of an "optimized" C and C++ codegen I raised elsewhere before - what does me babbling "optimized" mainly entail?

(not mutually-exclusive ascertainment proposals following:)

- less and simplified resulting code: easier to pick up/grasp for the C/C++ compiler for to do its own optimization heuristics, less parsing effort for us mere humans trying to debug, especially in the case of interop
- build time reduction: less code emission I/O, runtime string formatting for output...
- easier access for fresh contributors and better maintainability
- interop improvements
- further runtime optimizations

I am eagerly looking forward to the results of the LLVM-based undertakings, but I also think we can do a bit better (as outlined above) with our current C/C++ backends till those come to fruition.

**Long story short**: this PR here focuses on the C++ backend, augmenting the current codegen method of establishing "temporary" variables by using C++11's auto type deduction. The reasons for adopting an "Almost Always Auto" style have been collected [here ](https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/) for the C++ world. For those hopping between C++'s and Nim's realms, this change also results in a bit less code and less work for the codegen part (no redundant `getTypeDesc`s): no need to tell the C++ compiler the type it already knows of (in most cases).

### CODE REVIEW

**Comment by ASVIEST** [comment]
I'm not sure what `auto` should be used here. Yes, in c++ code, it allows you to improve code performance by avoiding silent implicit conversions, but in theory, the type should always be detected correctly in the code generated in Nim, so such conversions should not appear. Is it really necessary ?

**Comment by heterodoxic** [comment]
> I'm not sure what `auto` should be used here. Yes, in c++ code, it allows you to improve code performance by avoiding silent implicit conversions, but in theory, the type should always be detected correctly in the code generated in Nim, so such conversions should not appear. Is it really necessary ?

I am sorry, but I don't understand your reservation: nowhere have I claimed to bring about "more" correctness in type determination to the C++ layer (though, if you had taken the time to at least have a glimpse at https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/ I added as a reference above, one could argument it mostly works towards ensuring more correctness).

Neither did I mention anything about some "necessity" here: the current codegen works. I am trying to make a case for this change to reduce both Nim compiler's workload and to shrink the code size a bit while keeping functional equivalence, for the latter being determined by passing tests.

**Comment by ASVIEST** [comment]
I put it a little wrong, I meant I don't see a lot of advantages from auto here.
1. In terms of the readability of the generated code. There is an item in the article "(Un)readability?", there are arguments why this is not such a big problem: loss of meaning of types due to a bad type system of C(++), with types the code is less generic, which contradicts the concept of C++. This is true when it comes to C, but it is incorrect with respect to the generated code, firstly, there should be no problems with incorrect types because ideally Nim should correctly guess the types in the backend (see [a]), secondly, the generated code is essentially an intermediate representation that takes care of the implementation, everything related to the code is done in the sem phase of the compiler.
2. I'm not sure if this can happen, but `auto` before C++17 cannot always construct an object, because there was no guaranteed copy elison, for example:
``
auto tmp = std:atomic<int>(0)
``
does not compile to C++17. As far as I understand, this can be generated, for example, by a procedure that returns `std:atomic` (obtained via ffi) and is used several times.

[a] The nim implementation of obtaining a C++ type may contain errors that can lead to the very implicit conversion that is avoided by `auto`. Suggestion: because this may slow down compilation, so I think it should be a flag that will add `explicit` for procedures with `constructor` pragma, as well as add something like this code for procedures:
(only C++11 syntax)
```c++
void func(int val) {}

template <typename T>
void func(T)
{
    static_assert(false, "Implicit convertion detected.");
}
```
Various type-checking functions can be generated for tmp parameters. 
In the case of C++20, you can make the implementation a little better:
```c++
#include <concepts>

template <typename T>
concept explicit_int = std::same_as<T, int>;

void fun(explicit_int auto x) {
  explicit_int auto a = 999999999;
}
```
Such a check will help to identify the problem of guessing C++ types in Nim.

But that's just my opinion, maybe auto is really better

**Comment by jmgomez** [comment]
Im not opposed to the change but it definitely hurts readability when debugging. AFAIK that function it's only used when `locNone` meaning you dont have the type in the right side but a function call.

**Comment by heterodoxic** [comment]
> I put it a little wrong, I meant I don't see a lot of advantages from auto here.
> 
>     1. In terms of the readability of the generated code. There is an item in the article "(Un)readability?", there are arguments why this is not such a big problem: loss of meaning of types due to a bad type system of C(++), with types the code is less generic, which contradicts the concept of C++. This is true when it comes to C, but it is incorrect with respect to the generated code, firstly, there should be no problems with incorrect types because ideally Nim should correctly guess the types in the backend (see [a]), secondly, the generated code is essentially an intermediate representation that takes care of the implementation, everything related to the code is done in the sem phase of the compiler.
> 
>     2. I'm not sure if this can happen, but `auto` before C++17 cannot always construct an object, because there was no guaranteed copy elison, for example:
>        `auto tmp = std:atomic<int>(0)`
>        does not compile to C++17. As far as I understand, this can be generated, for example, by a procedure that returns `std:atomic` (obtained via ffi) and is used several times.
> 
> 
> [a] The nim implementation of obtaining a C++ type may contain errors that can lead to the very implicit conversion that is avoided by `auto`. Suggestion: because this may slow down compilation, so I think it should be a flag that will add `explicit` for procedures with `constructor` pragma, as well as add something like this code for procedures: (only C++11 syntax)
> 
> ```c++
> void func(int val) {}
> 
> template <typename T>
> void func(T)
> {
>     static_assert(false, "Implicit convertion detected.");
> }
> ```
> 
> Various type-checking functions can be generated for tmp parameters. In the case of C++20, you can make the implementation a little better:
> 
> ```c++
> #include <concepts>
> 
> template <typename T>
> concept explicit_int = std::same_as<T, int>;
> 
> void fun(explicit_int auto x) {
>   explicit_int auto a = 999999999;
> }
> ```
> 
> Such a check will help to identify the problem of guessing C++ types in Nim.
> 
> But that's just my opinion, maybe auto is really better

Sorry for getting back to you this late, "lots of other stuff going on", and I am still a bit puzzled by the reluctance here which seems mostly by opinions rather than falsifiable assumptions on **what this change is trying to accomplish and what (un)intended effects it might have on codegen and debuggability**.

First of all: thanks for opting for more concrete arguments now: were we to continue a "discussion" on this change on an abstract level that you first offered, I fear we could never find a chance to find any form of consent here.

Regarding your second point, as we currently target [C++11] (https://github.com/nim-lang/Nim/pull/21378#issuecomment-1445040472), that could be a concern. Indeed, in the case of deleted copy constructors and/or move assignment operators, `auto`**`&&`** would be required. Yet, this case is independent from using auto type deduction, i.e. it fails for the current codegen as well: https://godbolt.org/z/vPe943TTP 

Regarding your first argument, I understand that the question of **readability**, raised by @jmgomez as well, is far more important here. I partially agree: resorting to the compiler's type deduction everywhere might worsen the situation of getting a grasp on what types are actually interacting, we cannot merely reason in highly abstract terms (all the time). But let me reply with a counter question: do you really use nothing but a human debugger for the emitted code? With mangling (for structs) and macros (`N_LIB_PRIVATE`, `NIM_CONST`, `N_NIMCALL_PTR`, `SEQ_DECL_SIZE`, `N_NIMCALL` and many more) everywhere, can you actually easily immediately tell what's going on there? Why do languages implement type interference again, is it only about the typing effort? Those constructs serve to hide actual (platform-specific) implementations, encouraging you to code against interfaces instead, keeping the focus on what degree of abstraction seems relevant for you most of the time. Oh, and BTW: "foreign" constructs from other headers will be `#include`d as well, but I guess one could still cognitively cope with that total information around in one's mind at all times.
For us boringly-ordinary humans, it seems to me we resort to compiler frontend tools that helps us resolve all those parts of the puzzle, **where we need/want it**, especially for debugging/optimization purposes. I dare claiming we already use those helpers most of the time, even when not debugging. So yes, `auto` type deduction in the specific instance the proposed change aims for, adds to abstractions, while still retaining complete information via respective tools (which BTW is why I celebrate the renewed efforts put into projects like [VSCode-Nim](https://github.com/nim-lang/vscode-nim): many, many thanks for this, @jmgomez, @nickysn, and @ringabout!).

Which brings me to my last point and back to my very first words: please do not "pre-judge" this PR as a sort of way to bring about a drastic change to the codegen by rashly replacing everything with `auto` type deduction. Please have a look at the concrete and isolated code line change which applies to return types of functions where Nim already assumes to know what to return, as @jmgomez has pointed out. It does not even touch the many, many pitfalls dug up [here](https://www.youtube.com/watch?v=wQxj20X-tIU) or [here](https://www.youtube.com/watch?v=ZCGyvPDM0YY) for example. For possible further expansions of `auto` replacements, further considerations would be needed indeed, that would result in new PRs with respective assessments of its feasibility.

Replacing the concrete type by auto inference will not only result in spitting out a shorter type name on the left side of assignments most of the time, but will also remove the need of invoking `getTypeDesc`, which already operates on a type cache to shorten lookup times. But no time still seems less to me than "some" time. This is merely a micro-optimization I admit, but take a look at the C++ implementation of the `system` module for instance: you get less code output, less runtime formatting, no type lookups in over 100 cases in debug mode. And you still retain the type info by either examining the function's return type manually or having the frontend provide you with that information.

I hope my broad remarks have contributed rather to a better understanding than confusion. :slightly_smiling_face:

**Comment by Araq** [comment]
Pros: 

- Shorter code generated.
- The code is easier to type-check for a C++ backend.

Cons:

- Less redundancy in the code can hide errors.
